jgrapht-opt/src/main/java/org/jgrapht/opt/graph/sparse/CSRBooleanMatrix.java

```diff
@@ -0,0 +1,230 @@
+            if (column < 0 || column >= columns) {
+                throw new IllegalArgumentException("Entry at invalid column: " + column);
+            }
+            columnIndices[cIndex++] = column;
+
+            // count non-zero per row
+            int row = e.getFirst();
+            rowOffsets[row + 1]++;
+        }
+
+        // prefix sum
+        Arrays.parallelPrefix(rowOffsets, (x, y) -> x + y);
+    }
+
+    /**
+     * Get the number of columns of the matrix.
+     * 
+     * @return the number of columns
+     */
+    public int columns()
+    {
+        return columns;
+    }
+
+    /**
+     * Get the number of rows of the matrix.
+     * 
+     * @return the number of rows
+     */
+    public int rows()
+    {
+        return rowOffsets.length - 1;
+    }
+
+    /**
+     * Get the number of non-zero entries of a row.
+     * 
+     * @param row the row
+     * @return the number of non-zero entries of a row
+     */
+    public int nonZeros(int row)
+    {
+        assert row >= 0 && row < rowOffsets.length;
+
+        return rowOffsets[row + 1] - rowOffsets[row];
+    }
+
+    /**
+     * Get an iterator over the non-zero entries of a row.
+     * 
+     * @param row the row
+     * @return an iterator over the non-zero entries of a row
+     */
+    public Iterator<Integer> nonZerosPositionIterator(int row)
+    {
+        assert row >= 0 && row < rowOffsets.length;
+
+        return new NonZerosIterator(row);
+    }
+
+    /**
+     * Get the position of non-zero entries of a row as a set.
+     * 
+     * @param row the row
+     * @return the position of non-zero entries of a row as a set.
+     */
+    public Set<Integer> nonZerosSet(int row)
+    {
+        assert row >= 0 && row < rowOffsets.length;
+
+        return new NonZerosSet(row);
+    }
+
+    private class NonZerosSet
+        extends
+        AbstractSet<Integer>
+    {
+        private int row;
+
+        public NonZerosSet(int row)
+        {
+            this.row = row;
+        }
+
+        @Override
+        public Iterator<Integer> iterator()
+        {
+            return new NonZerosIterator(row);
+        }
+
+        @Override
+        public int size()
+        {
+            return rowOffsets[row + 1] - rowOffsets[row];
+        }
+    }
+
+    private class NonZerosIterator
+        implements
+        Iterator<Integer>
+    {
+        private int curPos;
+        private int toPos;
+
+        public NonZerosIterator(int row)
+        {
+            this.curPos = rowOffsets[row];
+            this.toPos = rowOffsets[row + 1];
+        }
+
+        @Override
+        public boolean hasNext()
+        {
+            return (curPos < toPos);
+        }
+
+        @Override
+        public Integer next()
+        {
+            if (!hasNext()) {
+                throw new NoSuchElementException();
+            }
+            return columnIndices[curPos++];
+        }
+
+    }
+
+    private class PairComparator
```

Comment:
This belongs with Pair instead.

Reply:
Not sure how to do this, as this is specialized (and we are not using C++).

Reply:
I guess it would require a ComparablePair alternative which restricts the types of its parameters to Comparable.